pkg/replica/hash.go

```diff
@@ -0,0 +1,71 @@
+package replica
+
+import (
+	"os"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+type SnapshotHashStatus struct {
+	lock sync.Mutex
+
+	replica *Replica
+
+	SnapshotNames []string
+
+	State    ProgressState
+	Progress int
+	Checksum string
+	Error    string
+}
+
+func NewSnapshotHash(snapshotNames []string) *SnapshotHashStatus {
+	return &SnapshotHashStatus{
+		SnapshotNames: snapshotNames,
+		State:         ProgressStateInProgress,
+	}
+}
+
+func (s *SnapshotHashStatus) OpenSnapshot(snapshotNames []string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	dir, err := os.Getwd()
+	if err != nil {
+		return errors.Wrapf(err, "cannot get working directory")
+	}
+
+	snapshotDiskNames := []string{}
+	for _, name := range snapshotNames {
+		snapshotDiskNames = append(snapshotDiskNames, GenerateSnapshotDiskName(name))
+	}
+
+	r, err := NewReadOnly(dir, snapshotDiskNames[0], snapshotDiskNames, nil)
```

Comment:
Do we need to construct the struct `replica` for one file read? Can we directly open the sparse file (as a raw file)? This makes the modifications in `replica.go` look a little bit cumbersome.

Reply:
In some cases, it read a set of disk files, e.g. a snapshot disk file and a system-generated file.

Reply:
a system-generated file?

Reply:
After discussing with Shou, just hash non system-generated snapshot disk files.